### ISSUE TITLE
Rule request: disallow spread when calling a method

### ISSUE THREAD

**Issue by BetterCallSky** [palantir/tslint#4004]
Related to https://github.com/webpack/webpack/issues/7548, but probably it won't be fixed in webpack, and I think we can create a new tslint rule.

A new ts lint rule should report an error only when calling a method with a spread operator.
```ts
Obj.foo(...args); // error
```

Bellow usage is ok

```ts
foo(...args);

const bar = (...args) => {
// code
}

const {foo, ...rest} = props;
const c = {...a, ...b};
```

**Comment by giladgray** [comment]
@BetterCallSky i don't see how `Obj.foo(...args)` is wrong but `foo(...args)` is fine?

**Comment by BetterCallSky** [comment]
@giladgray
From https://github.com/webpack/webpack/issues/7548

Orginal code
```ts
Rx.concat(...args)
```
is transpiled to
```ts
Rx.concat.apply(Rx, args);
//              ^^
//        Because of this the whole Rx lib need to be included
```

The whole `Rx` module is marked as a side effect and tree-shaking is disabled.

-----

```ts
concat(...args)
```
is transpiled to
```ts
concat.apply(null, args);
```
In this case there is no `Rx` object, and it doesn't affect tree-shaking.

**Comment by JoshuaKGoldberg** [comment]
> Calling `Rx.concat(args)` is effectively the same as calling `concat.call(Rx, args)`. It implies that the this context of `concat` is the whole of `Rx`.

...which is why it's preferable to `import { concat }` instead of `import * as Rx`, right?

https://www.learnrxjs.io/operators/combination/concat.html

> ```js
> import { concat } from 'rxjs/operators';
> import { of } from 'rxjs';
> ```

By importing the entire `Rx` object, you're already telling the bundler that you want to import the entire object. It's generally preferred to import just the parts you need.

Closing this issue as non-actionable on TSLint's side, but @BetterCallSky  if I'm horribly wrong please do comment here and we can keep discussing!

**Comment by BetterCallSky** [comment]
@JoshuaKGoldberg 

There is nothing wrong with using `import * as Rx` because tree-shaking will remove unused functions.  
You can read about tree shaking here https://webpack.js.org/guides/tree-shaking/.
  
You can check it in action here https://github.com/BetterCallSky/ts-webpack-treeshaking-bug
  
I am pretty sure it's possible to detect code like this in ts-lint `Rx.concat(...args)`

**Comment by JoshuaKGoldberg** [comment]
Yes, thank you, I'm familiar with tree shaking 😉

**Comment by JoshuaKGoldberg** [comment]
Yes, thank you, I'm familiar with tree shaking 😉 what I see as the question here is whether your specific edge case in tree shaking is common enough to justify a new rule in TSLint core.

Is what you're doing with `import * as Rx` standard procedure? The learnrxjs.io link seems to imply it's preferable to explicitly import only what you need. Other large libraries such as office-ui-fabric-react _(see their [README.md](https://github.com/OfficeDev/office-ui-fabric-react/))_ also recommend using something like `library-name/lib/section` or `library-name/section` and explicitly importing only what you need.

Also, just to confirm: how exactly would this rule behave? Would it be that if an object is imported via `import * as`, its members can't have `...` calls such as `Rx.concat(...args)`?

**Comment by JoshuaKGoldberg** [comment]
Note: per #4534, this issue will be closed in less than a month if no PR is sent to add the rule. If you _really_ need the rule, custom rules are always an option and can be maintained outside this repo!

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is being deprecated** and no longer accepting pull requests for new rules. See #4534. 😱

If you'd like to see this rule implemented, you have two choices:
* **Recommended:** Use [typescript-eslint](https://typescript-eslint.io) to lint your TypeScript code with ESLint ✅
* **Legacy:** Write it as a [custom TSLint rule](https://palantir.github.io/tslint/develop/custom-rules/) 🤷‍♂️

👋 It was a pleasure open sourcing with you!

_If you believe this message was posted here in error, please comment so we can re-open the issue!_

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
